compiler/rustc_ast_lowering/src/lib.rs

```diff
@@ -1360,67 +1340,85 @@ impl<'a, 'hir> LoweringContext<'a, 'hir> {
 
         let opaque_ty_def_id = self.local_def_id(opaque_ty_node_id);
 
-        let mut collected_lifetimes = FxHashMap::default();
+        let mut collected_lifetimes = Vec::new();
+        let mut new_remapping = FxHashMap::default();
+
         self.with_hir_id_owner(opaque_ty_node_id, |lctx| {
-            let hir_bounds = if origin == hir::OpaqueTyOrigin::TyAlias {
-                lower_bounds(lctx)
-            } else {
-                lctx.while_capturing_lifetimes(
+            if origin != hir::OpaqueTyOrigin::TyAlias {
+                let lifetimes_in_bounds =
+                    lifetime_collector::lifetimes_in_bounds(&lctx.resolver, bounds);
+                debug!(?lifetimes_in_bounds);
+
+                collected_lifetimes = lctx.create_lifetime_defs(
                     opaque_ty_def_id,
-                    &mut collected_lifetimes,
-                    lower_bounds,
-                )
+                    &lifetimes_in_bounds,
+                    &mut new_remapping,
+                );
             };
+            debug!(?new_remapping);
             debug!(?collected_lifetimes);
 
-            let lifetime_defs = lctx.arena.alloc_from_iter(collected_lifetimes.iter().map(
-                |(_, &(span, p_id, p_name, _))| {
-                    let hir_id = lctx.lower_node_id(p_id);
-                    debug_assert_ne!(lctx.opt_local_def_id(p_id), None);
+            lctx.with_remapping(new_remapping, |lctx| {
+                let hir_bounds = lctx.lower_param_bounds(bounds, itctx);
+
+                let lifetime_defs = lctx.arena.alloc_from_iter(collected_lifetimes.iter().map(
```

Comment:
what is this? explain! Ah, it's the HIR for the new `'a` and friends that will be added to the TAIT? Explain it in terms of the `TestReturn` TAIT I gave in the example.

I personally would move this above lowering the bounds, because it comes earlier in the syntax we are creating:

```rust
type TestReturn<'a, T, 'x>
```